libraries/AP_GPS/AP_GPS_VecNAV.cpp

```diff
@@ -0,0 +1,345 @@
+// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//
+// Driver to read VectorNAV protocol
+// Swee Warman sweewarman@gmail.com
+// ***
+
+#include "AP_GPS.h"
+#include "AP_GPS_VecNAV.h"
+#include <AP_HAL/Util.h>
+
+#include <AP_Common/AP_Common.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#define VECNAV_INITMSG \
+        "$VNWRG,06,0*XX\r\n"                      /* Disable ascii output from GPS */ \
+        "$VNWRG,75,2,80,29,FFFF,0FFF,07FF*XX\r\n" /* Enable Binary output - Common,GPS and INS group */
+
+
+extern const AP_HAL::HAL& hal;
+
+const char AP_GPS_VecNAV::_initialisation_blob[] = VECNAV_INITMSG;
+
+AP_GPS_VecNAV::AP_GPS_VecNAV(AP_GPS &_gps, AP_GPS::GPS_State &_state, AP_HAL::UARTDriver *_port) :
+    AP_GPS_Backend(_gps, _state, _port)
+{
+
+    gps.send_blob_start(state.instance, _initialisation_blob, sizeof(_initialisation_blob));
+}
+
+void AP_GPS_VecNAV::ExtractDataGroup1(uint8_t *payload, struct BinGroup1 *msg){
```

Comment:
These extraction functions are extremely messy. Is there any reason why a packed structure couldn't be used instead, so that this copying could be completely avoided? It would be much cleaner, as well as faster.